client/session.go

```diff
@@ -501,80 +542,113 @@ func (s *session) FetchAll(namespace string, ids []string, startInclusive, endEx
 
 	fetchBatchOpsByHostIdx = s.fetchBatchOpArrayArrayPool.Get()
 
-	// Defer cleanup
-	defer func() {
-		// Return fetch ops to pool
-		for _, ops := range fetchBatchOpsByHostIdx {
-			for _, op := range ops {
-				s.fetchBatchOpPool.Put(op)
-			}
-		}
-
-		// Return fetch ops array array to pool
-		s.fetchBatchOpArrayArrayPool.Put(fetchBatchOpsByHostIdx)
-	}()
-
 	s.RLock()
-	majority = s.majority
+	majority = int32(s.majority)
 	for idx := range ids {
 		idx := idx
 
 		var (
+			wgIsDone int32
 			results  []encoding.Iterator
-			enqueued int
+			enqueued int32
 			pending  int32
 			success  int32
 			firstErr error
 			errs     int32
 		)
 
 		wg.Add(1)
+		allCompletionFn := func() {
+			var reportErr error
+			errsLen := atomic.LoadInt32(&errs)
+			if errsLen > 0 {
+				resultErrLock.RLock()
+				reportErr = firstErr
```

Comment:
`firstErr` is the error that occurred for this read, this is needed as otherwise we wouldn't have a reference to any of the errors that occurred for this ID.  The main reason `reportErr` is introduced is so we can default it to nil and only acquire the read lock if we truly need to read the contents of `firstErr` - it's simply an optimization in reducing the need to call the read lock to read `firstErr`.